File: src/libes/gis/range.c
/**********************************************************************
 *
 *  G_read_range (name, mapset, range)
 *      char *name                   name of map
 *      char *mapset                 mapset that map belongs to
 *      struct Range *range          struct to hold range info
 *
 *  Reads the data range information associated with map layer "map"
 *  in mapset "mapset" 
 *
 *   returns:    1  if successful
 *               0  missing, range faked using largest category
 *              -1  on fail
 *
 *  note:   a warning message is printed if the file is missing or incorrect
 *
 **********************************************************************
 *
 *  G_write_range (name, range)
 *      char *name                  name of map
 *      struct Range *range         struct holding range info
 *
 *  Writes the range information associated with map layer "map"
 *
 *   returns:    0  if successful
 *              -1  on fail
 *
 **********************************************************************
 *
 * G_init_range (range)
 *      struct Range *range         struct for range info
 *
 * initializes range structure for call to G_update_range()
 *
 **********************************************************************
 *
 * G_update_range (cat, range)
 *    CELL cat                    cat to be factored into range
 *    struct Range *range         struct for range info
 **********************************************************************
 *
 * G_get_range_min_max (range, min, max)
 *    struct Range *range;
 *    CELL *min, *max;
 **********************************************************************/

#include "gis.h"

G_read_range (name, mapset, range)
    char *name, *mapset;
    struct Range *range;
{
    FILE *fd;
    long x[4];
    char buf[200];
    int i;
    CELL ncats;

    range->nmin = 0 ;
    range->nmax = 0 ;
    range->pmin = 0 ;
    range->pmax = 0 ;
    fd = NULL;

    sprintf (buf,"cell_misc/%s", name);
    if (G_find_file2 (buf, "range", mapset))
    {
	fd = G_fopen_old (buf, "range", mapset);
	if (!fd)
	    goto error;

	if (!fgets (buf, sizeof buf, fd))
	    goto error;
	x[0]=x[1]=x[2]=x[3]=0;
	i = sscanf (buf, "%ld%ld%ld%ld", &x[0], &x[1], &x[2], &x[3]);
	if (i <= 0)
	    goto error;

	while (i-- > 0)
	{
	    if (x[i] < 0)
	    {
		if (range->nmin == 0 || x[i] < range->nmin)
		    range->nmin = x[i];
		if (range->nmax == 0 || x[i] > range->nmax)
		    range->nmax = x[i];
	    }
	    else if (x[i] > 0)
	    {
		if (range->pmin == 0 || x[i] < range->pmin)
		    range->pmin = x[i];
		if (range->pmax == 0 || x[i] > range->pmax)
		    range->pmax = x[i];
	    }
	}
	fclose(fd) ;
	return 1;
    }
    else
    {
	ncats = G_number_of_cats (name, mapset);
	if (ncats < 0)
	    return -1;
	range->pmax = ncats;
	if(range->pmax > 0) range->pmin = 1;
	return 0;
    }

error:
    if (fd)
	fclose(fd) ;
    sprintf (buf, "can't read range file for [%s in %s]", name, mapset);
    G_warning (buf);
    return -1;
}

G_write_range (name, range)
    char *name;
    struct Range *range;
{
    FILE *fd;
    char buf[200];

    sprintf (buf,"cell_misc/%s", name);
    fd = G_fopen_new (buf, "range");
    if (!fd)
	goto error;

    fprintf (fd, "%ld %ld %ld %ld\n",
	(long)range->nmin, (long)range->nmax,
	(long)range->pmin, (long)range->pmax);
    fclose (fd);
    return 0;

error:
    sprintf (buf, "can't write range file for [%s in %s]",
	name, G_mapset());
    G_warning (buf);
    return -1;
}

G_update_range (cat, range)
    CELL cat;
    struct Range *range;
{
    if (cat > 0)
    {
	if (range->pmin == 0 || cat < range->pmin)
	    range->pmin = cat;
	if (range->pmax == 0 || cat > range->pmax)
	    range->pmax = cat;
    }
    else if (cat < 0)
    {
	if (range->nmin == 0 || cat < range->nmin)
	    range->nmin = cat;
	if (range->nmax == 0 || cat > range->nmax)
	    range->nmax = cat;
    }
}

G_row_update_range (cell, n, range)
    CELL *cell;
    struct Range *range;
{
    CELL cat;

    while (n-- > 0)
    {
	cat = *cell++;
	if (cat > 0)
	{
	    if (range->pmin == 0 || cat < range->pmin)
		range->pmin = cat;
	    if (range->pmax == 0 || cat > range->pmax)
		range->pmax = cat;
	}
	else if (cat < 0)
	{
	    if (range->nmin == 0 || cat < range->nmin)
		range->nmin = cat;
	    if (range->nmax == 0 || cat > range->nmax)
		range->nmax = cat;
	}
    }
}

G_init_range (range)
    struct Range *range;
{
    range->nmin = 0 ;
    range->nmax = 0 ;
    range->pmin = 0 ;
    range->pmax = 0 ;
}

G_get_range_min_max(range, min, max)
    struct Range *range;
    CELL *min, *max;
{
    *min = range->nmin ? range->nmin : range->pmin;
    *max = range->pmax ? range->pmax : range->nmax;
}

File: src.garden/grass.java/GRASS/Range.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public
    License along with this library; if not, write to the Free
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
    USA
*/
/*
 * Copyright (C) 1998 The Information-technology Promotion Agency, Japan(IPA), LGPL
 * $Id$
 */
package GRASS;

/**
 * This class wraps GRASS library struct Range. 
 *
 * See the <B>GRASS 4.2 Programmer's Manual</B>, 
 * Section 12.20.5 <I>struct Range</I>
 * 
 * @author  $Author$
 * @version $Revision$
 *
 * @see GRASS.libgis
 *
 */
public class Range {

  /**
   * Handles memory location of the actual struct Range.
   */
  private long addr = 0;

  /**
   * Constructs an empty object and allocates memory for GRASS struct Range.
   */
  public Range() {
    allocate();
    G_init_range();
  }
  
  // 12.10.5 Raster Range File
  /**
   * Reads and creates range object for the raster file in the specified
   * mapset. Wraps GRASS library routine:
   *
   * <pre>
   * G_read_range (name, mapset, range)
   * char *name;
   * char *mapset;
   * struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @param name GRASS raster file name
   *
   * @param mapset Name of mapset where raster file should reside
   *
   * @return Object GRASS.Range that contains range information of the
   *         specified raster file
   *
   * @exception GRASSException Exception with an appropriate error message 
   *            is thrown if some error occured while reading raster Range
   *            information.
   *
   */
  public Range(String name, String mapset) throws GRASSException {
    allocate();
    G_read_range(name, mapset);
  }
  
  /**
   * Used by other methods to allocate memory for GRASS struct Range.
   * Must be never called from outside of methods of this class.
   */
  protected native void allocate();
  
  /**
   * Reads the range information for the raster file name in mapset 
   * into the range structure. Wraps GRASS library routine:
   *
   * <pre>
   * G_read_range (name, mapset, range)
   * char *name;
   * char *mapset;
   * struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @param name GRASS raster file name
   *
   * @param mapset Name of mapset where raster file should reside
   *
   * @exception GRASSException Exception with an appropriate error message 
   *            is thrown if some error occured while reading raster Range
   *            information.
   */
  protected native void G_read_range(String name, String mapset)
    throws GRASSException;

  /**
   * Writes the range information for the raster file name in the current 
   * mapset from the range structure. Wraps GRASS library routine:
   *
   * <pre>
   * G_write_range (name, range) write raster range file
   *          char *name;
   *  struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @param name GRASS raster file name
   *
   * @exception GRASSException Exception with an appropriate error message 
   *            is thrown if some error occured while reading raster Range
   *            information.
   */
  public native void G_write_range(String name) throws GRASSException;
  // this method was never tested !!

  /**
   * Initializes the range structure for updates.
   * Wraps GRASS library routine:
   *
   * <pre>
   * G_init_range (range)
   *      struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   */
  public native void G_init_range();

  /**
   * Compares the cat value with the minimum and maximum values in
   * the range structure, modifying the range if cat extends the range.
   * Wraps GRASS library routine:
   *
   * <pre>
   * G_update_range (cat, range)
   *            CELL cat;
   *            struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @param cat category value
   *
   */
  public native void G_update_range(int cat);

  /**
   * This routine updates the range data just like G_update_range, 
   * but for values from the row of cell values.
   * Wraps GRASS library routine:
   *
   * <pre>
   * G_row_update_range (cell, n, range)
   *         CELL *cell;
   *         int n;
   *         struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @param buf buffer of cell values
   *
   * @param n number of values to be used for update
   *
   */
  public void G_row_update_range(CELL buf, int n) {
    
    G_row_update_range_( buf.getAddr(), n );

  }

  /**
   * The actual method to update the range data just like G_update_range, 
   * but for values from the row of cell values. Must be never called by
   * any methods other than memebers of this class. 
   * Wraps GRASS library routine:
   *
   * <pre>
   * G_row_update_range (cell, n, range)
   *         CELL *cell;
   *         int n;
   *         struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @param buf buffer of cell values
   *
   * @param n number of values to be used for update
   *
   */
  private native void G_row_update_range_(long addr, int n);

  /**
   * Extracts the minimum value from the range structure.
   * Wraps GRASS library routine:
   *
   * <pre>
   * G_get_range_min_max (range, min, max)
   *       struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @return the minimum value
   *
   */
  public native int min();

  /**
   * Extracts the maximum value from the range structure.
   * Wraps GRASS library routine:
   *
   * <pre>
   * G_get_range_min_max (range, min, max)
   *       struct Range *range;
   * </pre>
   *
   * See the <B>GRASS 4.2 Programmer's Manual</B>, 
   * Section 12.10.5 <I>Raster Range File</I>
   *
   * @return the maximum value
   *
   */
  public native int max();

  // Cleaning after
  /**
   * Used by other methods to free memory of GRASS struct Range.
   * Must be never called from outside of methods of this class.
   */
  protected native void free();
  
  /**
   * Deallocates memory for GRASS struct Range and calls finalize
   * method of the superclass.
   * Must be never called from outside of methods of this class.
   *
   * @exception Throwable if thrown by superclass finalizer
   */
  protected void finalize() throws Throwable {
    
    free(); //explicitly release memory
    super.finalize();

  }
  
  /**
   * Returns a string representation of the object.
   * 
   * @return String that represents this object 
   *
   */
  public String toString() {
    return getClass()+" at addr="+addr+" (min="+min()+", max="+max()+")";
  }

  
}
